Add documentation for the custom Jinja2 global function `undef` (#953)
ansible/ansible-documentation: docs/docsite/rst/playbook_guide/playbooks_templating_undef.rst

.. _templating_undef:

The undef function: add hint for undefined variables
====================================================

.. versionadded:: 2.12

The Jinja2 ``undef()`` function returns a Python ``AnsibleUndefined`` object, derived from ``jinja2.StrictUndefined``. Use ``undef()`` to undefine variables of :ref:`lesser precedence <ansible_variable_precedence>`.

For example, ``roles:`` defaults/vars are scoped to the play and a role could avoid inheriting another role's default by using ``undef()``.

.. code-block:: yaml

    roles:
      - role_a
      - role_b

.. code-block:: yaml

    # roles/role_a/defaults/main.yml
    role_default: role_a

    # roles/role_b/defaults/main.yml
    role_default: "{{ undef(hint='If this is required and missing, this error will be displayed. This must be defined with higher precedence than role defaults.') }}"

> .. seealso:: :ref:`DEFAULT_PRIVATE_ROLE_VARS` is a general way to do this for M(ansible.builtin.include_role)/M(ansible.builtin.import_role).

Reviewer: `M(...)` doesn't work here:
```suggestion
.. seealso:: :ref:`DEFAULT_PRIVATE_ROLE_VARS` is a general way to do this for :ansplugin:`ansible.builtin.include_role#module`/:ansplugin:`ansible.builtin.import_role#module`.
```

Author: Thanks, fixed

> The ``undef`` function accepts 1 optional argument:

Reviewer: ```suggestion
The ``undef`` function accepts one optional argument:
```
(https://docs.ansible.com/ansible/devel/dev_guide/style_guide/grammar_punctuation.html#numbers)

> hint

Reviewer: Shouldn't this be
```suggestion
:hint:
```
?

Author: I don't know, but I was keeping consistency with the syntax used here https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_templating_now.html.

Reviewer: Ah. With `:hint:` it would look more like this: https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-fields

For consistency let's keep your form then. I think it makes sense to switch over to `:...:` eventually, but that's better suited for another PR (which changes all related files at the same time).

Author: Thanks for the pointer.
